### PR TITLE
Adding Working Directory to Start-Job cmdlet

### LINKED ISSUE #4287: Start-Job needs a -WorkingDirectory parameter
Currently, `Start-Job`:

* defaults to _different_, _fixed_ working directories on _different platforms_ (which is problematic in itself):
   * `$HOME` on Unix (macOS, Linux)
   * `$HOME\Documents` on Windows

* by contrast, using the newly-introduced Unix-like `... &` syntax defaults to the _current_ location; this discrepancy is problematic too [_update: but as designed_] - see #4267

Either way, there is **no simple way to have the _caller_ set the working directory _explicitly_,** leading to such painful workarounds as in [this SO answer](https://stackoverflow.com/a/17388991/45375).

The proposed solution:

```powershell
# Wishful thinking
> $jb = Start-Job -WorkingDirectory $PSHOME { "Hi from $PWD." }; Receive-Job -AutoRemove -Wait $jb
Hi from C:\Program Files\PowerShell\6.0.0-beta.4
```

Environment data
----------------

```powershell
PowerShell Core v6.0.0-beta.4 on macOS 10.12.5
PowerShell Core v6.0.0-beta.4 on Ubuntu 16.04.2 LTS
PowerShell Core v6.0.0-beta.4 on Microsoft Windows 10 Pro (64-bit; v10.0.15063)
Windows PowerShell v5.1.15063.413 on Microsoft Windows 10 Pro (64-bit; v10.0.15063)
```

### PR DESCRIPTION
**Pull request by davinci26**
# PR Summary

1. Added `WorkingDirectory` to `PSRemotingCmdlet.cs`
2. Inherited and validated the `WorkingDirectory` in `StartJob.cs`
3. Added a `Set-Location` to the Pipeline of  `Commands`

Thinks I would like to get your feedback on:

1. Implementation & approach
2. Test coverage, let me know if you would like to see more thorough testing
3. Any edge case that I might have missed

Awaiting your feedback and comments!

## PR Context

Fix #4287 
## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: I will file an issue when the pr is in later stages
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs [hunk 1]
```diff
@@ -1188,6 +1188,25 @@ public virtual ScriptBlock ScriptBlock
...
+        /// <summary>
+        /// Initial location of the script block.
+        /// This is used to set the location prior to executing the script block
+        /// </summary>
+        public virtual string WorkingDirectory
+        {
+            get
+            {
+                return _scriptBlockInitialLocation;
+            }
```
Comment (vexx32) [inline src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: I don't think the explicit backing field here is necessary. You can define this like so:

```csharp
public virtual string WorkingDirectory { get; set; }
```

and then just reference `WorkingDirectory` instead of the private field.

Comment (vexx32) [inline src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: I don't think the explicit backing field here is necessary. You can define this like so:

```csharp
public virtual string WorkingDirectory { get; set; }
```

and then just reference `WorkingDirectory` instead of the private field.

File: src/System.Management.Automation/engine/remoting/commands/StartJob.cs [hunk 2]
```diff
@@ -480,6 +480,25 @@ public virtual ScriptBlock InitializationScript
...
+        { 
+            get
+            {
+                return base.WorkingDirectory;
+            }
+
+            set
+            {
+                base.WorkingDirectory = value;
+            }
```
Comment (vexx32) [inline src/System.Management.Automation/engine/remoting/commands/StartJob.cs]: This can be simplified to:
```csharp
get => base.WorkingDirectory;
set => base.WorkingDirectory = value;
```

File: src/System.Management.Automation/engine/remoting/commands/StartJob.cs [hunk 3]
```diff
@@ -623,24 +654,23 @@ protected override void CreateHelpersForSpecifiedComputerNames()
                             ErrorCategory.PermissionDenied,
                             Context.LanguageMode));
             }
-
+
```
Comment (vexx32) [inline src/System.Management.Automation/engine/remoting/commands/StartJob.cs]: Looks like there a handful of lines in this file you added a bit of whitespace that doesn't need to be here. Probably best to remove it so as not to complicate the diffs in any future edits. 🙂

File: test/powershell/engine/Job/Jobs.Tests.ps1 [hunk 4]
```diff
@@ -69,6 +69,14 @@ Describe 'Basic Job Tests' -Tags 'CI' {
             $ProgressMsg[0].StatusDescription | Should -BeExactly 2
         }
 
+        It 'Can use the user specified working directory parameter'{
+            $parent = [System.IO.Path]::GetTempPath()
+            $job = Start-Job -ScriptBlock { $pwd } -WorkingDirectory $parent| Wait-Job
```
Comment (vexx32) [inline test/powershell/engine/Job/Jobs.Tests.ps1]:
```suggestion
            $job = Start-Job -ScriptBlock { $pwd } -WorkingDirectory $parent | Wait-Job
```

File: test/powershell/engine/Job/Jobs.Tests.ps1 [hunk 5]
```diff
@@ -69,6 +69,14 @@ Describe 'Basic Job Tests' -Tags 'CI' {
...
+            Wait-Job $job
+            Receive-Job $job -OutVariable jobOutput
```
Comment (vexx32) [inline test/powershell/engine/Job/Jobs.Tests.ps1]: Is there a reason you wrote it this way? I would usually opt for something more straightforward, like...
```suggestion
            $jobOutput = Receive-Job $job
```

File: src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: Also doc comment must start with `Gets or sets`.

File: src/System.Management.Automation/engine/remoting/commands/StartJob.cs [hunk 6]
```diff
@@ -480,6 +480,25 @@ public virtual ScriptBlock InitializationScript
 
         private ScriptBlock _initScript;
 
+
+        /// <summary>
+        /// Working directory of the process.
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/commands/StartJob.cs]:
```suggestion
        /// Gets or sets a working directory of the process.
```

File: src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs [hunk 7]
```diff
@@ -1843,6 +1862,13 @@ internal Pipeline CreatePipeline(RemoteRunspace remoteRunspace)
 
             pipeline.Commands.Clear();
 
+            if (_scriptBlockInitialLocation != null)
+            {
+                var command = new Command("Set-Location");
+                command.Parameters.Add("LiteralPath", _scriptBlockInitialLocation);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: Here can be a security issue. 
Should we check that is really directory and it is impossible inject bad code?

/cc @PaulHigin @TravisEz13

**Comment by iSazonov** [comment]
@davinci26 Please look CodeFactor issues. They should be fixed for new code.

File: src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs (hunk 7 above)
Comment (davinci26) [reply src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: This is a really good point.

I do check the directory to see if its valid in `StartJob.cs` [here](https://github.com/davinci26/PowerShell/blob/837eb6a5b6a389a080b5685ea87a27bbb0021169/src/System.Management.Automation/engine/remoting/commands/StartJob.cs#L611) and throw if its not.

The reason I did not make the check in the `PSRemotingCmdlet` is that I wanted to enable the scenario where this parameter is expanded to other cmdlets that inherit from `PSRemotingCmdlet`. 

That being said I could be really defensive and add a command that checks that the string provided by the user is indeed a valid path.

File: src/System.Management.Automation/engine/remoting/commands/StartJob.cs (hunk 6 above)
Comment (davinci26) [reply src/System.Management.Automation/engine/remoting/commands/StartJob.cs]: Thanks for the review @iSazonov
fixed in https://github.com/davinci26/PowerShell/commit/396369a7fe660166cd829e4a99f63fd913760e9c

File: test/powershell/engine/Job/Jobs.Tests.ps1 (hunk 5 above)
Comment (davinci26) [reply test/powershell/engine/Job/Jobs.Tests.ps1]: @vexx32 thanks for the review
fixed in https://github.com/davinci26/PowerShell/commit/06ef81a76db4898ec1066f69e50558d47f8e85f0

File: test/powershell/engine/Job/Jobs.Tests.ps1 (hunk 4 above)
Comment (davinci26) [reply test/powershell/engine/Job/Jobs.Tests.ps1]: fixed in https://github.com/davinci26/PowerShell/commit/06ef81a76db4898ec1066f69e50558d47f8e85f0

File: src/System.Management.Automation/engine/remoting/commands/StartJob.cs (hunk 3 above)
Comment (davinci26) [reply src/System.Management.Automation/engine/remoting/commands/StartJob.cs]: fixed in https://github.com/davinci26/PowerShell/commit/06ef81a76db4898ec1066f69e50558d47f8e85f0

File: src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs (hunk 1 above)
Comment (davinci26) [reply src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: fixed in https://github.com/davinci26/PowerShell/commit/06ef81a76db4898ec1066f69e50558d47f8e85f0

File: src/System.Management.Automation/engine/remoting/commands/StartJob.cs (hunk 2 above)
Comment (davinci26) [reply src/System.Management.Automation/engine/remoting/commands/StartJob.cs]: No strong opinion here but: I am tempted to leave this as it is and the reason for that is the rest of the file that I am touching follows this pattern for the get\sets [source](https://github.com/davinci26/PowerShell/blob/837eb6a5b6a389a080b5685ea87a27bbb0021169/src/System.Management.Automation/engine/remoting/commands/StartJob.cs)

File: src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs [hunk 8]
```diff
@@ -1843,6 +1849,13 @@ internal Pipeline CreatePipeline(RemoteRunspace remoteRunspace)
 
             pipeline.Commands.Clear();
 
+            if (WorkingDirectory != null)
+            {
+                var command = new Command("Set-Location");
+                command.Parameters.Add("LiteralPath", WorkingDirectory);
+                pipeline.Commands.Add(command);
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: Something about this doesn't feel right. Starting with the test:

```PowerShell
Start-Job -ScriptBlock { Join-Path $pwd "" } -WorkingDirectory $tempPath 
```

I think this code does the equivalent of:

```PowerShell
Set-Location $tempPath | Join-Path $pwd ""
```

Instead of:

```PowerShell
Set-Location $tempPath
Join-Path $pwd ""
```

I think you need to call `PSCommand.AddStatement` on this command you're adding.

File: src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs [hunk 9]
```diff
@@ -1843,6 +1849,13 @@ internal Pipeline CreatePipeline(RemoteRunspace remoteRunspace)
 
             pipeline.Commands.Clear();
 
+            if (WorkingDirectory != null)
+            {
+                var command = new Command("Set-Location");
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: This should probably the module, like:

```C#
var command = new Command("Microsoft.PowerShell.Management\\Set-Location");
```

File: src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs (hunk 8 above)
Comment (davinci26) [reply src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: If I understand correctly what you are saying is that the object `Pipeline` represents only the powershell pipeline and if I want to create a sequence of commands I would need to add a statement to the `powershell` instance instead of adding a command to the pipeline.

This means that the correct approach is to have something like this:

``` C#
_powershellV3 = System.Management.Automation.PowerShell.Create();
if (WorkingDirectory != null)
{
    _powershellV3.AddStatement().
             AddCommand("Microsoft.PowerShell.Management\\Set-Location").
             AddParameter("LiteralPath", WorkingDirectory);
}
``` 

I would assume that this statement should be added for both when the server is ps2 and ps5+.

(P.S. thanks for the comment it helped me understand the code way better)

Comment (KirkMunro) [reply src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: I think what @lzybkr is trying to draw your attention to is what happens if `WorkingDirectory != null` on line 1852 of PSRemotingCmdlet.cs is true. In that case, you add the `Set-Location` command to the empty set of pipeline commands (the collection that was previously cleared on line 1850), but then logic continues beyond your if-statement and commands are added _as additional commands in that same pipeline that `Set-Location` is in_ on lines 1859-1862. If you insert `pipeline.Commands.AddStatement();` before line 1857, you'll properly terminate the statement that invokes `Set-Location`, allowing the commands that are being invoked to be property added as a separate commands following that statement.

File: src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs (hunk 7 above)
Comment (TravisEz13) [reply src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: Adding a command should be considered like parameterized SQL.  

Impossible is the wrong bar here.  

@iSazonov I'm not sure what the problem is with setting the working directory with a string parameter (this is not a scriptblock).

File: src/System.Management.Automation/engine/remoting/commands/StartJob.cs (hunk 2 above)
Comment (KirkMunro) [reply src/System.Management.Automation/engine/remoting/commands/StartJob.cs]: Generally speaking from what I've seen, when most contributors add new code, they do so using the current practices rather than match existing syntax. This keeps us forward moving. Eventually PRs come in to advance old syntax to catch up to newer syntax.

File: src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs (hunk 8 above)
Comment (lzybkr) [reply src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: The `PowerShell` api was originally designed for a single pipeline, but you could do more with that api **if** you used `AddScript`.

Some secure endpoints disallow script (no language mode), so the `PowerShell` instance must be expressible as commands. At some point in the past, this also meant a "Pipeline" was exactly that, it was a single pipeline.

I think in V3 we added support for multiple distinct statements via the `AddStatement` api. You can think of `AddCommand` as adding a `|` between every statement *unless* `AddStatement` was called, then we'd add a `;` instead. I also think that means the correct code is more like:

```c#
    _powershellV3.
             AddCommand("Microsoft.PowerShell.Management\\Set-Location").
             AddParameter("LiteralPath", WorkingDirectory).
             AddStatement();
```

I do see that the test passed though, and that surprised me. I'd do some more testing to verify my assumptions above.

I don't recall specifics about going between the PowerShell api and the Pipeline api, and to be honest, I find the original code here (before the PR) a little confusing, so I'm reluctant to say that `AddStatement` is definitely needed.

I should also mention the need for compatibility testing if `AddStatement` is needed. I recall a bug where the conversion from a script block to a PowerShell instance could not use `AddStatement` because of a bug in the server side code. If I'm remembering this correctly, the bug was fixed in V5, so it would probably be worth testing across different server versions.

**Comment by PaulHigin** [comment]
I don't think this is the right way to implement working directory.  I am wary of modifying any script run by the user ... if there is an error or unexpected behavior, then it becomes confusing for the user to understand what is wrong with their script.

I think a better way may be to set the working directory path when the job child process is created.  This can be done as a PowerShell (pwsh) argument e.g., 
`pwsh.exe -WorkingDirectory ...`.

Or it can be done in the dotNet StartInfo property when creating the child process.
`_startInfo.WorkingDirectory = ...`

Please take a look at PowerShellProcessInstance.cs source file.

File: src/System.Management.Automation/engine/remoting/commands/StartJob.cs (hunk 2 above)
Comment (davinci26) [reply src/System.Management.Automation/engine/remoting/commands/StartJob.cs]: Sure, I will fix it in the next iteration

**Comment by davinci26** [comment]
@PaulHigin @KirkMunro 

I have looked at the code you mention ( that was my original approach) but to my suprise it does not work. At least with the current implementation. I have already posted some of this on #4287 and I would love to get your input. I re-post it here for convience:

### Changing the working directory of the pwsh remote server:
1. The command line argument `workingdirectory` does not work when the pwsh process runs in server mode (-s flag). See [consolehost.cs](https://github.com/PowerShell/PowerShell/blob/3ca5a1bdb031eeb9453b1481bcfa23eee6d89aec/src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs#L204). The function that initiates the process into server mode does not even take into consideration the working directory parameter provided by the user.

2. Using `_startInfo.WorkingDirectory = ...` also did not work either. The surprising part here ( at least for me) was that even though the new process working directory was the one specified by the process `_startInfo` the initial location did not change. I used process explorer [source](https://docs.microsoft.com/en-us/sysinternals/downloads/process-explorer) to verify that the process working directory was passed correctly.

### Thoughts & Questions

1. Since the InitializationScript parameter of the cmdlet is passed to the process startupInfo as part of the process arguments, I do not see how we could have a workingDirectory parameter for Start-Job that would be able to set the working directory also for the InitializationScript. Unless we enable the `workingDirectory` parameter in `serverMode`. Is this part of the requirement?

2. Do you think that enabling the `workingDirectory` parameter when powershell runs in server mode is a good addition?

3. Is there documentation (or some old issue that I could read to understand better) on the communication protocol between PowerShell and Powershell server mode? 

Let me know what you think :)

**Comment by PaulHigin** [comment]
Ok, well I can't say I understand how dotNet StartInfo works, but it seems like the WorkingDirectory parameter is ignored in some (non ShellExecute) cases.  PowerShell does not seem to set initial working directory location except for the case where the -WorkingDirectory command parameter is used.

The only alternative I see for Start-Job is to pass the   -WorkingDirectory argument to the Run method that starts PowerShell running in server mode.

https://github.com/PowerShell/PowerShell/blob/d84ebc157196e80c9ddb116f351bf633fd1340e7/src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs#L204
```
System.Management.Automation.Remoting.Server.OutOfProcessMediator.Run(s_cpp.InitialCommand, s_cpp.WorkingDirectory);
```

And then pass it on through to server session creation until it reaches the ServerRunspacePool constructor, similar to configurationName argument.  Then in the runspace creation handler, run the `Set-Location` command before running initialization scripts.

https://github.com/PowerShell/PowerShell/blob/cf7699b4b9bb96bac906a7052f890d524f375677/src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs#L619

**Comment by lzybkr** [comment]
@PaulHigin - Maybe this [blog post](https://devblogs.microsoft.com/oldnewthing/20071211-00/?p=24223) answers your question, assuming the process is running elevated.

**Comment by PaulHigin** [comment]
@lzybkr  Ok, that makes sense to protect against dll planting.  So again I think the solution for Start-Job as I outlined above is the right solution, since it basically implements -WorkingDirectory for a job child process.

**Comment by davinci26** [comment]
@PaulHigin  @iSazonov @lzybkr @KirkMunro 

I am following the approach that Paul mentioned above. Since the implementation is completely different Id rather not have them merged in to the same git history and a big portion of the discussion here is implementation specific.

That is why I will do the following:
1.  I will abandon this PR in favor of new one
2. I will update the issue to include a summary of the comments above.

See the PR #10324
